Fix non-table filtering issue

bprusinowski opened the pull request:
I have not considered all possible situations in #191 – I assumed that we don't need to differentiate between tables and other charts when dealing with filters, but in fact we do.

This PR fixes an issue currently available on DEV, when using a given dimension as a segment,  which removes currently applied single filter, and this behavior is not desired in this case.

I have also added some comments, to not forget in the future what's going on here (along with another test) 😆

herrstucki commented on app/configurator/configurator-state.tsx:
```diff
@@ -261,7 +272,7 @@ export const applyDimensionToFilters = (
       // Add filter for this dim if it's not one of the selected multi filter fields, but
       // only when the dimension isn't grouped, as only then we need to switch to a single filter.
-      if (isHidden && !isGrouped) {
+      if ((isHidden || !isField) && !isGrouped) {
```
Nit: I find this condition quite hard to understand 😅 Is there a way to make it more clear or at least update the comment to describe the cases that are covered? But maybe it's just me 👴

bprusinowski replied:
I'll try to! 👨‍💻

herrstucki commented on app/configurator/configurator-state.tsx:
```diff
@@ -261,7 +272,7 @@ export const applyDimensionToFilters = (
       // Add filter for this dim if it's not one of the selected multi filter fields, but
       // only when the dimension isn't grouped, as only then we need to switch to a single filter.
-      if (isHidden && !isGrouped) {
+      if ((isHidden || !isField) && !isGrouped) {
```
Nit: I find this condition quite hard to understand 😅 Is there a way to make it more clear or at least update the comment to describe the cases that are covered? But maybe it's just me 👴

herrstucki commented on app/configurator/configurator-state.spec.tsx:
```diff
@@ -111,13 +111,26 @@ describe("deriveFiltersFromField", () => {
+    isField,
+    isHidden,
+    isGrouped,
+  }: {
```
Thank you for naming these options! Much clearer now than before!

bprusinowski commented:
Based on the fact that both of you mentioned that it's not very easy to understand what's going on there (and it also took me a while to get this) I have refactored the code and added the comments, so hopefully it'll be easier now.

herrstucki commented on app/configurator/configurator-state.tsx:
```diff
@@ -202,82 +202,147 @@ export const getFilterValue = (
-    );
+    const { chartType, fields, filters } = chartConfig;
+
+    if (chartType === "table") {
```
Yes, this makes it much easier to understand 👌

herrstucki commented on app/configurator/configurator-state.tsx:
```diff
@@ -202,82 +202,147 @@ export const getFilterValue = (
-      }
+        break;
+      default:
+        throw new Error();
```
Is this here to check that this case should in fact never happen? If so, I recommend using TypeScript to do the work for us instead of throwing at runtime: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking

bprusinowski replied:
Yes - never used it previously, thanks!

herrstucki commented on app/configurator/configurator-state.spec.tsx:
```diff
@@ -1,5 +1,8 @@
+import {
+  applyNonTableDimensionToFilters,
+  applyTableDimensionToFilters,
+} from ".";
```
Nit: shouldn't this be `from "./configurator-state"?

bprusinowski replied:
Right, auto-imports 😅

herrstucki commented on app/configurator/configurator-state.tsx:
```diff
@@ -202,82 +202,147 @@ export const getFilterValue = (
+        }
+        break;
+      default:
+        throw new Error();
```
See above comment about exhaustiveness checking

herrstucki reviewed (approved):
Thanks! Looks much clearer to me now!

ptbrowne reviewed (approved):
Thanks, it is indeed much clearer now !

ptbrowne commented on app/configurator/configurator-state.spec.tsx:
```diff
@@ -126,11 +128,11 @@ describe("deriveFiltersFromField", () => {
     const expectedFiltersState = {};
 
-    _applyDimensionToFilters(initialFiltersState, false, false);
+    _applyDimensionToFilters(initialFiltersState, true, false, false);
```
Ah I forgot to post these comment yesterday. Nevermind...

ptbrowne commented on app/configurator/configurator-state.spec.tsx:
```diff
@@ -194,7 +196,25 @@ describe("deriveFiltersFromField", () => {
+    expect(initialFiltersState).toEqual(expectedFiltersState);
+  });
+
+  it("non-table: should not modify filters if dimension is not used as a field", () => {
```
Also forgot to post it yesterday, this is absolutely not blocking.
